scipy/optimize/_basinhopping.py

```diff
@@ -25,7 +25,7 @@ def _add(self, minres):
         self.minres.x = np.copy(minres.x)
 
     def update(self, minres):
-        if minres.fun < self.minres.fun:
+        if minres.success and minres.fun < self.minres.fun:
```

Comment:
Why this change? If the function, for any reason, is smaller than the current global minimum, it's a better minimum, isn't it? It might not be the global minimum if the minimisation failed, but it's certainly a better than a local minimum with a higher function value?

Reply:
Especially if a minimiser is being used with a small `nfev`, many minimisations will fail, but basinhopping can still be smart choice for finding approximations to the global minimum. I would argue that it should at least be an option to save as global minimum even if minimisation failed.

Reply:
@juliusbierk  The reason I started making these changes was that it was keeping results even when minimization failed, leading to results that violated the given constraints: https://github.com/scipy/scipy/issues/7799

Reply:
@endolith Ah! I see. Yes, with bounds I understand why this is a problem. I see other use-cases, however, where the former behaviour is preferred (for my use cases, that's typically the case). Could this simply be made an option? Or perhaps bounds should be passed directly to `basinhopping`? Not sure at all what the best approach, and maybe I'm fairly alone with the preference for the former behaviour.

The values that minimizers return will always be the result of a function evaluation, won't it? At least with the minimizers I know of.

Reply:
Well I thought of passing bounds or constraints to basinhopping itself, but it still needs to pass them to the local minimizer, and it would also need some kind of tolerance argument too, or it will reject a lot of things because of floating point inequality, like the local minimizer will find 49.99999998 as meeting a constraint of >=50, which the basinhopping would then reject.

Reply:
I see. Yes, that would make it a lot more tedious. 

I agree with you. It won't be a real minimum, but it is the "lowest value found", and I would always expect the lowest value found to be returned rather the "the lowest minimum found".

Reply:
Yes I agree that it should keep "the lowest value found" as long as it meets the bounds and constraints, and is the result of an evaluation of the original function, even if the minimization fails.  But how do we do that, since going out of constraints causes a minimization failure, too.